Trials part 2 - review (#3)
ferrous-systems/unsize-experiments: README.md

It defines how the unsizing of the inner type occurs for a given pointer or wrapper type.

A `CoerceUnsized` implementation has specific requirements to be valid which boil down to 2 kinds:
1. A non-delegating `CoerceUnsized` impl
2. A delegating `CoerceUnsized` impl

### 1. A non-delegating `CoerceUnsized` impl

Such an impl is used for actual pointer like types, such as `&'a T` or `Arc<T>`.
The implementing type and the `CoerceUnsized` target type must differ in a single generic parameter only. Say, the parameters are `T` and `U`. Then,

> - `T` is the generic paramter of the implementing type; is bound as `T: Unsize<U>`

Reviewer: ```suggestion
- `T` is the generic parameter of the implementing type; is bound as `T: Unsize<U>`
```

> - `U` is the generic paramter of the `CorecedUnsized` target type

Reviewer: ```suggestion
- `U` is the generic parameter of the `CoerceUnsized` target type
```

```rust
// SAFETY:
// - `Unsize::target_metadata` returns length metadata that spans the entire array exactly.
// - `[T; N]` is a contiguous slice of `T`'s, so a pointer pointing to its data is valid to
//   be interpreted as a pointer to a slice `[T]`.
unsafe impl<T, const N: usize> Unsize<[T]> for [T; N] {
    fn target_metadata((): <Self as Pointee>::Metadata) -> <[T] as Pointee>::Metadata {
        N
    }
}
```

> ### `CoercedUnsized`

Reviewer: ```suggestion
### `CoerceUnsized`
```

// SAFETY:
// - `Unsize::target_metadata` returns length metadata that spans the entire array exactly.
// - `[T; N]` is a contiguous slice of `T`'s, so a pointer pointing to its data is valid to
//   be interpreted as a pointer to a slice `[T]`.
unsafe impl<T, const N: usize> Unsize<[T]> for [T; N] {
    fn target_metadata((): <Self as Pointee>::Metadata) -> <[T] as Pointee>::Metadata {
        N
    }
}
```

### `CoerceUnsized`

The non-delegating implementations of `CoerceUnsized` provided by the standard library will have the implementation of their `fn coerce_unsized` function written to disassemble the source into pointer and source metadata, make use of the `Unsize` trait for extracting the target metadata from the source metadata, and then reassembling the pointer and target metadata into the target.

For the delegating implementations, the implementation of the `fn coerce_unsized` function will merely delegate to the inner value and then wrap that result again.

## Implementations provided by the compiler

> ⚠️ Note: This section uses fictional rust syntax

The compiler will generate `Unsize` implementations for types to trait object for their implemented types as before:

> For types to trait object for their implemented types, the compiler will generate `Unsize` implmentations:

Reviewer 2: Now this is duplicated

Author: jesus i made this mistake

        Pin::new(self.pointer.coerce_unsized())
    }
}

// Permit going from `Pin<impl Pin>` to `Pin<impl Pin>`
impl<P, U> CoerceUnsized<Pin<U>> for Pin<P>
where
    P: CoerceUnsized<U>,
    P: core::ops::Deref<Target: !Unpin>,
    U: core::ops::Deref<Target: !Unpin>,
{
    fn coerce_unsized(self) -> Pin<U> {
        // SAFETY: The new unpin Pin is derived from another unpin Pin,
        // so we the pinned contract is kept up
        unsafe { Pin::new_unchecked(self.pointer.coerce_unsized()) }
    }
}
```

While this is a breaking change, it should be in line with being a soundness fix.

> Unfortunately, these kind of impl requires negative bounds and negative reasoning which is its own can of worms and therefore likely not to happen, see [GH issue: Need negative trait bound #42721](https://github.com/rust-lang/rust/issues/42721).

Reviewer 2: I always get this wrong 🤦‍♂️
